cou/apps/core.py

```diff
@@ -106,7 +106,44 @@ def post_upgrade_steps(
         :return: List of post upgrade steps.
         :rtype: list[PostUpgradeStep]
         """
-        return self._get_enable_scheduler_step(units) + super().post_upgrade_steps(target, units)
+        return (
+            self._get_enable_scheduler_step(units)
+            + self._get_restart_subordinate_services_steps(units)
+            + super().post_upgrade_steps(target, units)
+        )
+
+    def _get_restart_subordinate_services_steps(
+        self, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Get step to restart all subordinate services if they aren't running.
+
+        :param units: Units to restart subordinate services.
+        :type units: Optional[list[Unit]]
+        :return: Steps to restart to subordinate services
+        :rtype: list[PostUpgradeStep]
+        """
+        if not units:
+            units = list(self.units.values())
+        steps = []
+        for unit in units:
+            for subordinate in unit.subordinates:
+                if "ceilometer-agent" == subordinate.charm:
+                    service = "ceilometer-agent-compute"
+                    steps.append(
+                        PostUpgradeStep(
+                            description=(
+                                f"Restart subordinate service for unit: '{subordinate.name}'"
+                            ),
+                            coro=self.model.run_on_unit(
+                                unit_name=subordinate.name,
+                                command=(
+                                    f"systemctl is-active --quiet {service}"
+                                    f" || systemctl restart {service}"
+                                ),
+                            ),
```

Comment:
I think that this step could be simpler, saying: "Restart subordinate service for ceilometer-agent-compute units".
Like this we will not need to have `SubordinateUnit` and the coroutine here will simply run over all (if NovaCompute is run with units, all here will mean all subordinate to those units) subordinates units of ceilometer-agent-compute.

Like this, we lose information about specific unit in step, but for operator such information is not really relevant I think. It must properly say that it will restart only related units. The implementation will be simpler.

Reply:
I see your point. The complexity your proposal want to avoid is because the data transforming layer in COU, which I 100% agree with you it is a hard decision to change it.

So let me explain my:

- I would like to make each restart service as a single step, the plans is more clear to me.
- The implementation I gave doesn't require more effort to get the data from the controller.
- This SubordinatesUnit design provide more flexible possible when we need to upgrade the hypervisor.

I wouldn't say which one is better, just trade-off.

I will let other people provide their feedback.

Reply:
Your points are correct, and I'm not against them, I just try to avoid COU being more complex.

Reply:
Hi @rgildein 

Since no other provided their feedback. I will try to pitch my idea:

There is a major differences between our proposal:

I want to show the clear plan, every service restart should have a single COU step. Without the `SubordinateUnit`, the possible way will be checking if the service exists in the unit. This require more networking effort at both planning and upgrading.

(If I misunderstanding your proposal, please correct me)

Reply:
non-blocking, just a comment: for me, I would simple add a post upgrade step to restart ceilometer agent-compute service in `NovaCompute` class only, since we know that service will be in `NovaCompute` unit,

Reply:
> since we know that service will be in NovaCompute unit

The `ceilometer-agent-compute` service exists only if `ceilometer-agent` is subordinate to nova-compute. You can not know this information without using SubordinateUnit or collecting more information from the unit.

Reply:
check exit status of `systemctl list-unit-files ceilometer-agent-compute`

Reply:
my opinion here is that the implementation of this `SubordinateUnit`indeed brings a little bit more complexity, but it opens more doors if in the future we need to be smarter about other subordinate units. Will this happen? I don't think it will, but it's already implemented and it's more connected with what libjuju does.

Reply:
> The ceilometer-agent-compute service exists only if ceilometer-agent is subordinate to nova-compute. You wouldn't know this information without using SubordinateUnit or collecting more information from the unit.

You can simply check if such charm is subordinate to it or not. Step will check if there is any and if yes restart it. Operator will see the description saying: "Check subordinate ceilometer-agent and restart services ".

Reply:
After discussion, we agreed that a simpler solution would be more appropriate, but since it's already done, it's good to go.